7.x: new router and middleware system

hekike opened the pull request:
Replace the router and middleware logic with a more performant solution.
Goal: zero or almost zero API/functionality change with significant performance improvement.

Pt 1. (https://github.com/restify/node-restify/pull/1554) was nice for dumb servers and synthetic tests; this should be more useful for real-world applications, especially ones with many routes.

# Changes

- Replaces the router with the Radix Tree based [find-my-way](https://github.com/delvedor/find-my-way)
- New middleware system called `chain.js`, similar to [connect](https://github.com/senchalabs/connect)
- Move versioning and contentType routing into a `conditionalHandler` plugin, see:

```js
server.use(restify.plugins.conditionalHandler({
   contentType: 'application/json',
   version: '1.0.0'
   handler: function (req, res, next) {
       next();
   })
});

server.get('/hello/:name', restify.plugins.conditionalHandler([
  {
     version: '1.0.0',
     handler: function(req, res, next) { res.send('1.x') }
  },
  {
     version: ['1.5.0', '2.0.0'],
     handler: function(req, res, next) { res.send('1.5.x, 2.x') }
  },
  {
     version: '3.0.0',
     contentType: ['text/html', 'text/html']
     handler: function(req, res, next) { res.send('3.x, text') }
  },
  {
     version: '3.0.0',
     contentType: 'application/json'
     handler: function(req, res, next) { res.send('3.x, json') }
  }
]);
```

Hints for review:

- `response.js`: 100% new, do not review it by diff
- `server.js`: the complete request handling logic is new, I recommend to review it as a new file

# TODO

- [x] replacing router
- [x] replacing middleware system
- [x] content-type based routing *(moved to `conditioalHandler` plugin)*
- [x] router versioning, see: https://github.com/delvedor/find-my-way/issues/39 *(moved to `conditioalHandler` plugin)*
- [x] fix plugin tests *(114 passing,  39 failing)*
- [x] ~~revisit fix response tests port *(why did it fail with absolut url?)*~~
- [x] add back DTrace support to server.js
- [x] `router` tests
- [x] `chain` tests
- [x] re-visit `once` for handler `next`
- [x] router unmount, see: https://github.com/delvedor/find-my-way/issues/40 *(tests are disabled temporary)*
- [x] Write migration guide

# Breaking Changes

- Nonstrict-routing is not available
- Limited `RegExp` usage in router path see https://github.com/delvedor/find-my-way
- Remove already deprecated `next.ifError`
- disable DTrace probes by default *( I know it's painful, but most of the people don't know/use it, it causes ~10% throughput improvement in all benchmark tests)*, `createServer({ dtrace: true })`
- Router versioning and contentType routing moved to `conditioalHandler` plugin, see above
- `once ` is not applied automatically for multiple `next()` calls, use the new `onceNext: true` server option, `strictNext` is the same

# Results

Test machine: Mid 2014 MBP, 2.6 i5 16GB RAM 
(~30% improvement compared to `6.3.2` (https://github.com/restify/node-restify/pull/1554), ~50% compared to `6.3.1`)

## Result: dumb server

*Note: Most of the performance improvements in these benchmarks come from the new middleware logic and other optimizations, the router performance is not tested here as it uses LRU cache.*

```
---- response-json ----
✔ Results saved for head/response-json
✔ Results saved for stable/response-json
response-json throughput:
{
    "significant": "***",
    "equal": false,
    "wins": "head",
    "diff": "30.21%"
}

---- response-text ----
✔ Results saved for head/response-text
✔ Results saved for stable/response-text
response-text throughput:
{
    "significant": "***",
    "equal": false,
    "wins": "head",
    "diff": "34.18%"
}
```

## Result: middleware

```
---- middleware ----
✔ Results saved for head/middleware
✔ Results saved for stable/middleware
middleware throughput:
{
    "significant": "***",
    "equal": false,
    "wins": "head",
    "diff": "44.68%"
}
```

## Result: heavy router

*Note: Disabling (mocking out) cache in a stable version: it's not fair as it aims to the worst case, cache hit ratio is 0%. However, it's still better than the worst case scenario as it least doesn't require extra time to maintain the LRU cache.
There is no other way to simulate 100+ different endpoint calls with the current benchmark suite.*

```
---- router-heavy ----
✔ Results saved for head/router-heavy
✔ Results saved for stable/router-heavy
router-heavy throughput:
{
    "significant": "***",
    "equal": false,
    "wins": "head",
    "diff": "3086.8%"
}
```

hekike commented:
@DonutEspresso @retrohacker @yunong what do you think, should I continue? Even if I'm going to end up with necessary API change?

retrohacker commented on lib/middle.js:
```diff
@@ -0,0 +1,258 @@
+var proto = {};
+
+/* istanbul ignore next */
+var defer =
```
Why are we checking for the existence of `setImmediate`?

hekike replied:
Sry, I didn't make it clear that this is `connect`, I copy-pasted it, because I needed to extend it with the `next(false) logic. :(  Later we can try to open a PR or replace `next(false)` with something else.

hekike commented:
@retrohacker I added two more benchmarks: middleware and router-heavy
You can find the results in the description.

hekike commented:
Question: should `next(false)` only prevent to run the handlers in the same category or all of the handlers, which one:

1. `next(false)` in pre, stops the next `pre` handlers but runs all the `use` and router handlers
2. `next(false)` in pre, stops the next `pre` and `use` handlers but runs router
3. `next(false)` in pre, stops all the upcoming handlers

hekike commented:
Question: I don't find where the `createServer({ dtrace })` options it handled, does it still exist? I mean this option itself?

DonutEspresso commented:
Whoaaaaaaa this is awesome 🎉 🎉 🎉  

> Question: should next(false) only prevent to run the handlers in the same category or all of the handlers

The intention of false was to originally stop all subsequent middlewares in the event you had flushed a response already and considered this request complete. So I think 3 should be the desired behavior (if it isn't already).

hekike commented:
@DonutEspresso @yunong @retrohacker  I have an issue with the 'close' and 'aborted' test cases:
https://github.com/restify/node-restify/blob/master/test/server.test.js#L2648

No matter how I end/destroy/close the req/res/socket I always have both the `server.on('aborted')` and `server.on('close')` events, I cannot abort without `close` event, and I cannot close without `aborted` event. Is it possible to trigger `close` event without `aborted`? Does it make sense to separate the two connection states in `restify`? If I understand correctly we will always have `close` https://nodejs.org/api/http.html#http_event_aborted

yunong commented:
@hekike I think it may be the case that Node core actually fires both events? We could check the source code to be sure, but if that's the case, then it's okay that both events fire.

DonutEspresso commented:
Hmm great question - I was under impression a close wouldn't cause an abort but I can see both events being fired now. Short answer: I don't know! May have to dig into core to see if there is anything worth differentiating there.

hekike commented:
**close and aborted:** Thanks for the help! I've checked out Node core and seems to be that `emit('aborted')` is always immediately followed by an `emit('close')` in node core. *(I'm not 100% sure)* I also opened a question here: https://github.com/nodejs/help/issues/978

See: [_http_server.js#L417](https://github.com/nodejs/node/blob/3eb5778d1b1408d6223cdb3f5124a9b83f46f5eb/lib/_http_server.js#L417) and [http2/compat.js#L14](https://github.com/nodejs/node/blob/11a9f36cae0d1e6e11c9f40ec82d9f887aa0a911/lib/internal/http2/compat.js#L142)

in 7.x I unified server `close` and `aborted` events. restify throws only `RequestCloseError`, see: https://github.com/restify/node-restify/pull/1561/commits/0da821920f735a1b9b194d0b3ddcde360a0f77f3

DonutEspresso commented on lib/server.js:
```diff
@@ -779,420 +700,250 @@ Server.prototype.toString = function toString() {
+ * @fires Request,Response#request
  */
-Server.prototype._routeAndRun = function _routeAndRun(req, res) {
+Server.prototype._onRequest = function _onRequest(req, res) {
```
So clean, so simple 💯

DonutEspresso commented on lib/server.js:
```diff
@@ -1222,7 +980,122 @@ Server.prototype._setupRequest = function _setupRequest(req, res) {
+
+/**
+ * Wrapper method for emitting the after event. this is needed in scenarios
+ * where the async formatter has failed, and the ot assume that the
```
This comment looks a little outdated since we don't have async formatters anymore - don't want to bikeshed on this PR but could make a note to come back and clean up comments in another PR.

DonutEspresso commented on lib/server.js:
```diff
@@ -1528,7 +1283,7 @@ function errEvtNameFromError(err) {
         return 'VersionNotAllowed';
     } else {
-        return err.name.replace(/Error$/, '');
+        return err.name ? err.name.replace(/Error$/, '') : 'Error';
```
Under what scenarios does `err.name` not exist?

DonutEspresso commented on lib/server.js:
```diff
@@ -779,420 +700,250 @@ Server.prototype.toString = function toString() {
+//            4.2 _afterRoute (runs after last handler called next)
+//              4.3 _afterRoute (when, next('string') called)
+//              4.3 _routeErrorResponse (when error happaned in router)
+// 5. _finishReqResCycle (on response "finish" and "error" events)
```
This waterfall is great! The additional granularity is nice. Is this order different than the current behavior? I thought the order is as follows:

1. pre
2. router.find
3. use (use only applies to routes that were successfully routed)
4. done/done(err)

hekike replied:
You are right, I was dumb enough to work based on my `express` and `koa` experience instead of reading the docs carefully :/

> Allows you to add in handlers that run for all routes. Note that handlers added via use() will run only after the router has found a matching route. If no match is found, these handlers will never run. - restify docs

*(In express/koa everything is a middleware, the router as well, so they don't make difference between pre/after route, it depends on the order of the middleware "registration")*

Btw maybe it's misleading for others as well coming from the `express` world, shouldn't we keep `use` as an alias and introduce a new term that reflects it's behavior better? Like `afterRoute()` or `useRoute()`, these are not very good names, but something like these :)

DonutEspresso commented on benchmark/benchmarks/router-heavy.js:
```diff
@@ -0,0 +1,132 @@
+    : require('restify');
+
+var server = restify.createServer();
+var path = '/whiskeys/scotch/islay/lagavulin/16-years/50';
```
🍺

hekike commented on lib/chain.js:
```diff
@@ -0,0 +1,185 @@
+ * @public
+ * @class Chain
+ * @param {Object} [options] - options
+ * @param {Boolean} [options.onceNext=false] - Prevents calling next multiple
```
Yes, it's for backward compatibility. Making `strictNext` to the default causes performance degradation.

hekike commented:
@DonutEspresso I fixed the `use`, `route` handler running order issue. As a result of this, I also reverted all the related changes. Please check out the breaking changes list again, as it's smaller now.

I tried to improve the flow by using `vasync` and it's "competitors" but I always experienced significant performance degradation. I would skip it for now. Btw https://github.com/suguru03/neo-async was the fastest.

If I understand correctly then only the conditional handler need be improved to get your approval.

hekike commented:
@DonutEspresso with control flow: https://gist.github.com/hekike/26d486b2321c802513613fba86ca5b99

It's like 5-10% slower on synthetic tests, would you prefer this format?

hekike commented:
I updated this PR with the latest master, moved the current Radix tree based router lookup logic to a separate component with a more general interface. It will provide a way to use different router logic in the future.

Benchmarks on the updated PR with my new machine (pretty much the same) :p

cc @DonutEspresso @yunong 

```
pmarton@nflx node-restify (feat/new-router)*$ make benchmark 
up to date in 0.377s
? Do you want to track progress? No
? Do you want to compare HEAD with the stable release (6.3.4)? Yes
? Do you want to run all benchmark tests? Yes
? How many connections do you need? 100
? How many pipelining do you need? 10
? How long does it take? 30
---- response-json ----
✔ Results saved for head/response-json
✔ Results saved for stable/response-json
response-json throughput:
{
    "significant": "***",
    "equal": false,
    "wins": "head",
    "diff": "32.7%"
}

---- response-text ----
✔ Results saved for head/response-text
✔ Results saved for stable/response-text
response-text throughput:
{
    "significant": "***",
    "equal": false,
    "wins": "head",
    "diff": "36.39%"
}

---- router-heavy ----
✔ Results saved for head/router-heavy
✔ Results saved for stable/router-heavy
router-heavy throughput:
{
    "significant": "***",
    "equal": false,
    "wins": "head",
    "diff": "2702.78%"
}

---- middleware ----
✔ Results saved for head/middleware
✔ Results saved for stable/middleware
middleware throughput:
{
    "significant": "***",
    "equal": false,
    "wins": "head",
    "diff": "43.48%"
}
```

DonutEspresso commented on docs/guides/6to7guide.md:
```diff
@@ -0,0 +1,157 @@
+
+## Breaking Changes
+
+### Server throws `RequestCloseError` instead of `RequestAbortedError`
```
I don't think the server actually "throws" right? Do we mean that the server just logs a `RequestCloseError`? We may also want to update RequestCloseError's default status code to 444.

DonutEspresso commented on docs/guides/6to7guide.md:
```diff
@@ -0,0 +1,157 @@
+DTrace probes comes with some performance impact that's fine for the sake of
+observability but you may don't use it at all.
+
+### Change in calling `next` multiple times
```
Is performance the motivation behind this new default behavior?

retrohacker replied:
Also have a preference for this to remain unchanged. Throwing by default helps us catch a whole class of bugs.

hekike replied:
Yes, performance is the motivation behind.
@retrohacker do you mean at Netflix? We can re-enable it with the option for NQ services.

retrohacker replied:
@hekike I just mean in general. We are a framework for building production ready REST services, I'm not sure this is in line with our goals as a project.

DonutEspresso commented on docs/guides/6to7guide.md:
```diff
@@ -0,0 +1,157 @@
+### Router versioning and content type
+
+`accept-version` and `accept` based conditional routing moved to the
+`conditioalHandler` plugin, see docs or example:
```
s/conditioalHandler/conditionalHandler

DonutEspresso commented on lib/errorTypes.js:
```diff
@@ -3,4 +3,4 @@
 
 errors.makeConstructor('RequestCloseError');
-errors.makeConstructor('RequestAbortedError');
+errors.makeConstructor('RouteMissingError');
```
Where is this error used?

hekike replied:
This error existed before but was created locally.
When a code calls `next('route-name')` but the route doesn't exist.
It's a programmatic error.

DonutEspresso commented on lib/plugins/inflightRequestThrottle.js:
```diff
@@ -65,8 +65,7 @@ function inflightRequestThrottle(opts) {
             );
-            res.send(plugin._err);
-            return next(false);
+            return next(plugin._err);
```
I can't comment on the exact line, but I think this file here needs to create the error inline to get a proper stack trace (rather than creating it ahead of time).

DonutEspresso commented on lib/request.js:
```diff
@@ -807,32 +809,47 @@ function patch(Request) {
+     * @function closed
+     * @returns {Boolean} is closed
+     */
+    Request.prototype.closed = function closed() {
```
👍

DonutEspresso commented on lib/chain.js:
```diff
@@ -0,0 +1,179 @@
+    this.onceNext = !!options.onceNext;
+    this.strictNext = !!options.strictNext;
+
+    // strictNext next enfoces onceNext
```
s/enfoces/enforces

DonutEspresso commented on lib/chain.js:
```diff
@@ -0,0 +1,179 @@
+    if (typeof handler.run === 'function') {
+        var chain = handler;
+        handler = function nestedChainHandler(req, res, next) {
+            chain.run(req, res, next);
```
By chaining via a wrapper handler vs flattening the array, is there any impact on the result of `req.timers`?

hekike replied:
I removed the nested chain support, it's not used anywhere and not documented neither.

DonutEspresso commented on lib/chain.js:
```diff
@@ -0,0 +1,179 @@
+ * @instance
+ * @returns {Number} number of handlers in the stack
+ */
+Chain.prototype.count = function count() {
```
Potential bikeshed - why count() vs length()/.length property? (don't feel super strongly)

hekike replied:
Me neither, I thought it's a good idea to differentiate it from the JavaScript Array length

DonutEspresso commented on lib/chain.js:
```diff
@@ -0,0 +1,179 @@
+ * Invoke a handler.
+ *
+ * @private
+ * @param {Function} handler - rr function
```
`rr` function?

DonutEspresso commented on lib/router.js:
```diff
@@ -155,534 +27,351 @@ function compileURL(options) {
-        PUT: []
-    };
+    // Internals
+    this._anonymusHandlerCounter = 0;
```
s/anonymus/anonymous

DonutEspresso commented on lib/router.js:
```diff
@@ -155,534 +27,351 @@ function compileURL(options) {
-        this.reverse[route.path.source] = [];
-    }
+    handlers.forEach(function forEach(handler) {
+        // Assign name to anonymus functions
```
s/anonymus/anonymous

DonutEspresso commented on lib/router.js:
```diff
@@ -155,534 +27,351 @@ function compileURL(options) {
-    }
+    handlers.forEach(function forEach(handler) {
+        // Assign name to anonymus functions
+        handler._name =
```
Not sure I love that we're mutating user parameters - but wouldn't consider it blocking. Maybe worth adding a TODO for later (or filing issue).

DonutEspresso commented on lib/server.js:
```diff
@@ -83,13 +79,17 @@ var PROXY_EVENTS = [
- * @param {Boolean} [options.strictRouting=false] - If set, Restify
- * will treat "/foo" and "/foo/" as different paths.
+ * @param {Boolean} [options.noWriteContinue=false] - prevents
+ *  `res.writeContinue()` in `server.on('checkContinue')` when proxing
```
proxying?

hekike replied:
Hmm, I don't remember where I found this. I just documented it, could you help me with the correct behaviour?

DonutEspresso commented on lib/server.js:
```diff
@@ -496,8 +499,9 @@ Server.prototype.use = function use() {
-    argumentsToChain(handlers).forEach(function forEach(h) {
-        self.chain.push(h);
+    argumentsToChain(handlers).forEach(function forEach(handler) {
+        handler._name = handler.name || 'use-' + self.useChain.count();
```
Isn't this already happening in router?

hekike replied:
Router, pre, and use are registered separately with a different prefix.

DonutEspresso commented on lib/server.js:
```diff
@@ -466,8 +468,9 @@ Server.prototype.pre = function pre() {
-    argumentsToChain(handlers).forEach(function forEach(h) {
-        self.before.push(h);
+    argumentsToChain(handlers).forEach(function forEach(handler) {
+        handler._name = handler.name || 'pre-' + self.preChain.count();
```
Isn't this already happening in router?

hekike replied:
Router, pre, and use are registered separately with a different prefix.

DonutEspresso commented on lib/server.js:
```diff
@@ -779,420 +733,306 @@ Server.prototype.toString = function toString() {
+//         3.2. _afterRoute (runs after route handlers are finised,
+//                          triggers use)
+//          4.1 _runUse
+//            4.2. _afterUse
```
I know we chatted about this before, are we changing the pre -> use -> route order? Looks like this is breaking from 6.x.

hekike replied:
No, it's the same, the doc was wrong, fixed.

DonutEspresso commented on lib/server.js:
```diff
@@ -779,420 +733,306 @@ Server.prototype.toString = function toString() {
-
-    req.timers = [];
+    // Preserve handler err for finish event
+    res.err = res.err || err;
```
I didn't see where this is initially set - what's the use case for storing it on the response directly?

hekike replied:
It's used when the request succeeds but the response is an error:
https://github.com/restify/node-restify/blob/feat/new-router/lib/server.js#L1093

This is how it's ensured that error is always emitted in `after`.
It's not an elegant solution decorating the `res` object, but in the earlier version it was way to complex, I like that now it's based on the response events.

DonutEspresso commented on lib/server.js:
```diff
@@ -1223,7 +1070,122 @@ Server.prototype._setupRequest = function _setupRequest(req, res) {
+        res.removeListener('finish', onFinish);
+        self._finishReqResCycle(req, res, err);
+    }
+    res.once('finish', onFinish);
```
Is this a breaking change from 6.x? IIRC in 6.x we consider a request finished once it has reached the end of the handler chain. Using this pattern, it won't fire until the request is completely done. 

I have to think about whether this is better/worse/different, but definitely has an impact on how we might view metrics. cc @yunong

DonutEspresso commented on lib/server.js:
```diff
@@ -1223,7 +1070,122 @@ Server.prototype._setupRequest = function _setupRequest(req, res) {
+
+    // after event has signature of function(req, res, route, err) {...}
+    if (!res.finished && !req.closed()) {
+        res.once('finish', function resFinished() {
```
Since `finishReqResCycle()` doesn't get triggered until finish event fires, is this scenario possible anymore?

DonutEspresso commented:
Great work! Love that we added benchmarks - thanks for digging in and getting those perf numbers. 💯   Left some comments - mostly minor, the two big things for me are:

1) 6.x order is pre->use->route, order here looks like pre->route->use. Did we agree to a breaking change on that?
2) In 6.x a request is considered "complete" (i.e., causing audit log and metrics plugin to fire) when reaching the end of the next chain. "Completion" is synonymous with the 'after' event, where audit log and metrics are triggered. With this PR, "completion" gets delayed until the request object's finish event is fired. Not sure if this is good/bad yet but it is definitely different and may have an impact on how we interpret metrics. Thoughts from the rest of the team? cc @yunong @retrohacker

retrohacker commented on lib/chain.js:
```diff
@@ -0,0 +1,185 @@
+    }
+
+    // continue
+    next(error, req, res);
```
:+1: on leaving style changes to their own PR and fixing there if they pass lint. Suspect this isn't the only place in the code base we do this.

hekike commented:
We discussed the request "complete" topic in person with @DonutEspresso and we decided that the new behavior is acceptable and even better in some use cases like load shedding calculations where you really want to know when the request doesn't use your resources anymore. 

To have the correct timing about handler categories like `pre`, `use` and `router` execution durations a new timing object will be introduced in the next version of restify, related issue: https://github.com/restify/node-restify/issues/1603

hekike commented:
@DonutEspresso I added the new request timings and explained the change in the migration guide in a separate commit.

hekike commented:
We discussed in person that we need two latencies one when the request is flushed and one when all handlers are finished. The `after` event and inflight request counter will wait for both request flushing and handler finish.

DonutEspresso commented on lib/plugins/metrics.js:
```diff
@@ -39,7 +39,7 @@ function createMetrics(opts, callback) {
             method: req.method,
             // overall request latency
-            latency: hrTimeDurationInMs(req._time, process.hrtime()),
+            latency: hrTimeDurationInMs(req._time, req._timeFlushed),
```
Minus this last timer I think we are good to go!

hekike commented:
@DonutEspresso `totalLatency` added! https://github.com/restify/node-restify/pull/1561/commits/1acc1be51ce634eb571584581e6f2b960abe4a5b
